src/main/scala/util/Hierarchy.scala

```diff
@@ -0,0 +1,110 @@
+// See LICENSE for license details.
+
+package sifive.blocks.util
+
+import chisel3._
+import chisel3.util._
+
+import freechips.rocketchip.config._
+import freechips.rocketchip.diplomaticobjectmodel.logicaltree.LogicalTreeNode
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.subsystem._
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.prci._
+import freechips.rocketchip.util.LocationMap
+
+import firrtl.graph._
+
+case object HierarchyKey extends Field[Option[DiGraph[HierarchicalLocation]]](None)
+
+case class DevicesSubsystemParams(
+  name: String,
+  logicalTreeNode: LogicalTreeNode,
+  asyncClockGroupsNode: ClockGroupEphemeralNode)
+
+class DevicesSubsystem(val location: HierarchicalLocation, val ibus: InterruptBusWrapper, params: DevicesSubsystemParams)(implicit p: Parameters)
+  extends LazyModule 
+    with Attachable
+    with HasConfigurableTLNetworkTopology 
+    with CanHaveDevices {
+
+  def devicesSubhierarchies = None
+  def logicalTreeNode = params.logicalTreeNode
+  implicit val asyncClockGroupsNode = params.asyncClockGroupsNode
+
+  lazy val module = new LazyModuleImp(this) {
+    override def desiredName: String = params.name
+  }
+}
+
+trait HasConfigurableHierarchy { this: Attachable =>
+  def location: HierarchicalLocation
+
+  def createHierarchyMap(
+    root: HierarchicalLocation,
+    graph: DiGraph[HierarchicalLocation],
+    context: Attachable): Unit = {
+
+    hierarchyMap += (root -> context)
+
+    // Create and recurse on child hierarchies
+    val edges = graph.getEdges(root)
+    edges.foreach { edge =>
+      val dssParams = DevicesSubsystemParams(
+        name = edge.name,
+        logicalTreeNode = this.logicalTreeNode,
+        asyncClockGroupsNode = this.asyncClockGroupsNode)
+      val dss = context { LazyModule(new DevicesSubsystem(edge, ibus, dssParams)) }
+      createHierarchyMap(edge, graph, dss)
```

Comment:
It doesn't seem right that you are recursing into `dss` to be the next `context` but at the same time `DevicesSubsystem` sets `def devicesSubhierarchies = None`. Can you actually make a tree that is deeper than one level this way?

Reply:
Yeah, I understand how the name `devicesSubhierarchies` is confusing. You can make a tree that's arbitrarily deep this way, but the only place that cares about `deviceSubhierarchies` is the root of the hierarchy graph, which uses `hierarchyMap` to get references to all the subhierarchies; this is required for compatibility with the overlay API, which requires Seqs of each device type

Since this is the only thing that `hierarchyMap` is used for currently, maybe it would be more clear to remove it, populate each `DeviceSubsystem`'s `deviceSubhierarchies` in a way that makes sense, and then create the device Seqs the overlays need from that instead

Reply:
I think my rephrasing of this idea would be to have `DevicesSubsystem`s have `hierarchyMaps` which may or not not be populated with anything depending on their position in the tree?

Reply:
Yeah that would be doable, but if we did that today `DevicesSubsystem`s wouldn't have any use for their hierarchyMap, and that's a lot of redundant information if the root of the hierarchy tree's hierarchyMap contains the whole thing anyways...

Reply:
Well you said this approach was more bottom up so I think that is source of my confusion ;)
To me bottom up implies populating the leaf things first and then filling in the roots by collecting the information that has already been put in the leaves (so yes some redundancy, but is it harmful?). If the DeviceSubsystems can have DeviceSubsystems in them, it seems to me like their HierarchyMap should reflect that fact.

Reply:
Yeah, this is more bottom-up with the instantiation of the devices inside DevicesSubsystem, but the approach to connecting the subsystems is necessarily top-down to be compliant with the TLNetworkTopologyLocated connection API, which is why hierarchyMap is required (DevicesSubsystems don't have a hierarchyMap right now, only the trait which instantiates them does)